# PandAnnotate
A python-based alternative to Trinotate for merging herterogeneous transcriptome assembly annotation tables

Integrating information across annotations and nucleotide/protein database searches is a key step to understanding the functional content of a de novo transcriptome assembly, identifying contigs that may come from exogenous sources (e.g. bacteria, viruses, parasites, symbionts), establishing criteria to remove contigs from an assembly, and for performing many downstream analyses. The developers of Trinity created Trinotate, which adds annotations to a SQLite database, that is then written to, wait,....an Excel file. While Trinotate is relatively easy to use, it compresses many fields of a database search into a single column, delimited by heterogeneous separators that are difficult to parse, let alone load into R. Furthermore, it lacks flexibility by requiring the use of sequence-databases customized for Trinotate, not allowing integration of blastn searches, or other arbitrarily structured annotation information such as expression data. The goal of PandAnnotate is to provide a more flexible, pythonic implementation of database/search merges, and to write the merged database to a text file that is easy to load into R for further querying and analysis.
